### DISCUSSION TITLE
Problem converting to RINEX Obs/Nav

### DISCUSSION THREAD

**Discussion by alainmuls** [semuconsulting/pyrtcm#40] [Q&A] [answered]
Dear All,

I am working on converting RTCMv3 stream to RINEX observation and navigation files.

I work with Septentrio receivers and logged simultaneous the RTCMv3 and the SBF raw data. From both I cangenerate the observation and navigation data. The Obs/Nav RINEX files obtained by converting the SBF raw data are reference for getting the same values from the RTCM parsing I am creating.

1. For the observation conversion, I  have a single problem: the values I get for the PhaseRangeRate are in absolute value the same as those I have in the SBF converted RINEX file , except with the opposite sign. 

2. For the navigation files I still have some unresolved issues:
- the GPS week number does not take into account the GPS week rollover. Where can I get somewhere the date / time out of the RTCM messages
- for some datafields I get erroneous values (PS; I know about the SC to rad conversion needed). For example the DF079 (IDOT) when read gives the value 0 but should be 1.07147320259E-11, which is the value in the SBF converted RINEX Nav file and which I also get when I manually parse the 14 bits that are in the raw_data of RTCM, nl '00000000011110'. Similar errors are in DF083 (IDOT) and DF097 (deltan) and DF077 (SV ACC)
-  for some other datafields I get corresponding figures but with rounding errors in DF084 (af0) DF088 (M0) DF089 (Cuc) and several others

Something useful would be to be able to get the bits of a datafield, but I cannot find a suitable method.

Tx for any help

**Comment by alainmuls** [comment, marked as answer]
Hi All,

I found some corrections that needs to be done to get the same values as from a raw data SBF file:
1. values are rounded by  
SCALEDP = 8  # number of decimal places to scale float values to
which I changed to a larger value (24)

2. the procedure "bits2val" is not treating correctly the values. I changed it into:
def bits2valAM(att: str, scale: float, bitfield: int) -> object:
    """
    Convert bitfield to value for given attribute type.

    :param str att: attribute type e.g. "UNT008"
    :param float scale: scaling factor (where defined)
    :param int bitfield: attribute as integer
    :return: value
    :rtype: object (int, float, char, bool)
    """

    typ = rh.atttyp(att)
    siz = rh.attsiz(att)
    val = msb = 0

    if typ in ("SNT", "INT"):
        msb = 2 ** (siz - 1)
    if typ == "SNT":  # int, MSB indicates sign
        val = bitfield & msb - 1
        if bitfield & msb:
            val *= -1
    else:  # all other types
        val = bitfield
        # print(f"bitfield = {bitfield} :: {type(bitfield)}")
        # print(f"msb = {msb} :: {type(msb)}")
        # print(f"bitfield & msb = {bitfield & msb} :: {type(bitfield & msb)}")

    # if typ == "INT" and (bitfield & msb):  # 2's compliment -ve int
    #     val = val - (1 << siz)
    # if typ in ("CHA", "UTF"):  # ASCII or UTF-8 character
    #     val = chr(val)
    # # apply any scaling factor
    # else:
    if scale not in (0, 1):
        # val = round(val * scale, SCALEDP)
        val = val * scale

    return val

and with these modifications I get the same values as in the RINEX files

bye/Alain

**Comment by semuadmin** [comment]
Hi @alainmuls 

Thanks for this - you just beat me to it on the resolution. The SCALEDP value of 8 was chosen arbitrarily and I realise now that it was a bit too conservative for many applications. I'll test your proposed changes myself and if all's well get this into the next release of pyrtcm (v1.0.12).

**Comment by semuadmin** [comment]
Hi @alainmuls 

There's an updated version in branch https://github.com/semuconsulting/pyrtcm/tree/RC-1.0.12.

If you'd care to install it to confirm your earlier test results, be my guest:

```shell
pip install -U git+https://github.com/semuconsulting/pyrtcm.git@RC-1.0.12
```

**Reply by alainmuls** [reply]
Hi,

In my second post from yesterday the text that is very bold is normally commented out.

I tried to install the RC)1.0.12 but I get the following error:


[amuls:/tmp] $ pip install -U [git+https://github.com/semuconsulting/pyrtcm.git@RC-1.0.12](mailto:git+https://github.com/semuconsulting/pyrtcm.git@RC-1.0.12)
Defaulting to user installation because normal site-packages is not writeable
Collecting [git+https://github.com/semuconsulting/pyrtcm.git@RC-1.0.12](mailto:git+https://github.com/semuconsulting/pyrtcm.git@RC-1.0.12)
  Cloning https://github.com/semuconsulting/pyrtcm.git (to revision RC-1.0.12) to ./pip-req-build-w60lddew
  Running command git clone --filter=blob:none --quiet https://github.com/semuconsulting/pyrtcm.git /tmp/pip-req-build-w60lddew
  WARNING: Did not find branch or tag 'RC-1.0.12', assuming revision or ref.
  Running command git checkout -q RC-1.0.12
  error: pathspec 'RC-1.0.12' did not match any file(s) known to git
  error: subprocess-exited-with-error
 
  × git checkout -q RC-1.0.12 did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
 
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× git checkout -q RC-1.0.12 did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

Also the https link is not functional.

bye/alain

**Comment by semuadmin** [comment]
Resolved in https://github.com/semuconsulting/pyrtcm/commit/e0078d4a8b09fbce505388a0f0f44c4b26bf0983

**Reply by alainmuls** [reply]
Hi,

when I get access to the new release candidate I'll test it and let you know what the result is.

Alain

**Comment by semuadmin** [comment]
> Hi,
> 
> In my second post from yesterday the text that is very bold is normally commented out.

Not clear what your intentions were but the only change I've made is to remove the rounding. If you have any further code changes to submit, can I request that you submit them via a PR (Pull Request), so the changes are absolutely clear.

> I tried to install the RC)1.0.12 but I get the following error:

Release 1.0.12 is now generally available so the RC branch has been deleted.

**Reply by alainmuls** [reply]
Hi,

so just an update of pyrtcm is needed?

**Comment by semuadmin** [comment]
Yes

python3 -m pip install –upgrade pyrtcm

From: Alain Muls ***@***.***>
Date: Saturday, 5 August 2023 at 16:18
To: semuconsulting/pyrtcm ***@***.***>
Cc: semuadmin ***@***.***>, State change ***@***.***>
Subject: Re: [semuconsulting/pyrtcm] Problem converting to RINEX Obs/Nav (Discussion #40)

Hi,

so just an update of pyrtcm is needed?
